README.md
# react-soft-slider

![npm (tag)](https://img.shields.io/npm/v/react-soft-slider) ![npm bundle size](https://img.shields.io/bundlephobia/minzip/react-soft-slider) ![GitHub](https://img.shields.io/github/license/dbismut/react-soft-slider)

<p align="middle">
  <a href="https://nwj6597vq4.codesandbox.io"><img src="https://i.imgur.com/Lh8l5tB.gif" width="655"/></a>
</p>
<p align="middle">
  <a href="https://nwj6597vq4.codesandbox.io">Demo</a>
    <a href="https://codesandbox.io/s/nwj6597vq4"><b>[Source]</b></a>
</p>

React-soft-slider is a minimally-featured carousel. It focuses on providing the best user experience for manipulating slides. It doesn't try to implement additional features such as pagination dots, next and previous buttons, autoplay. If you're looking for a slider that has all this, there's [plenty](https://github.com/akiran/react-slick) of [alternatives](https://github.com/FormidableLabs/nuka-carousel) [out](https://github.com/express-labs/pure-react-carousel) [there](https://github.com/voronianski/react-swipe).

This allows react-soft-slider to be highly impartial when it comes to styling, so you shouldn't be fighting too hard to making the slider slider look the way you want.

- **Touch-gesture compatible:** handles swipe and drag on mobile and desktop devices
- **Spring animations:** driven by high-performance springs
- **Impartial styling:** you are responsible for the styling of your slides
- **Fully responsive:** as long as your slides styling is responsive as well!
- **Dynamic number of slides:** you can add or remove slides on the fly

React-soft-slider is powered by [react-spring](https://github.com/react-spring/react-spring) for springs animation and [react-use-gesture](https://github.com/react-spring/react-use-gesture) for handling the drag gesture.

## Installation

```
npm install react-soft-slider
```

> ⚠️ You also want to add the [intersection-observer](https://www.npmjs.com/package/intersection-observer) and [resize-observer](resize-observer-polyfill) polyfills for full browser support. Check out adding the [polyfills](#polyfills) for details about how you can include it.

## Usage

`<Slider />` has a very limited logic, and essentially does two things:

1. it positions the slider to the slide matching the `index` you passed as a prop
2. when the user changes the slide, it will then fire `onIndexChange` that will pass you the new `index`. You will usually respond by updating the slider `index` prop:

```jsx
import { Slider } from 'react-soft-slider'

const slides = ['red', 'blue', 'yellow', 'orange']
const style = { width: 300, height: '100%', margin: '0 10px' }

function App() {
  const [index, setIndex] = React.useState(0)

  return (
    <Slider
      index={index}
      onIndexChange={setIndex}
      style={{ width: 400, height: 200 }}
    >
      {slides.map((color, i) => (
        <div key={i} style={{ ...style, background: color }} />
      ))}
    </Slider>
  )
}
```

As you can see from the example, any child of the `<Slider />` component is considered as a slide. You are fully responsible for the appearance of the slides, and each slide can be styled independently.

> **Note:** although the above example uses hooks, react-soft-slider is compatible with Class-based components. However, since it internally uses hooks, it requires React `16.8+`.

### Props

The `<Slider />` component accepts the following props:

| Name              | Type                                | Description                                                                                                                                                                                                                                | Default Value                     |
| ----------------- | ----------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ | --------------------------------- |
| `children`        | `node`                              | elements you should pass to the slider and that will be considered as slides                                                                                                                                                               | Required                          |
| `index`           | `number`                            | the index of the slide that should be shown by the slider                                                                                                                                                                                  | Required                          |
| `onIndexChange()` | `(newIndex: number) => void`        | function called by the slider when the slide index should change                                                                                                                                                                           | Required                          |
| `indexRange`      | `[number,number]`                   | sets the minimum and maximum index range the slider should slide through. If the maximum index is negative, then it's set relatively to the children length. [See example here](https://codesandbox.io/s/react-soft-slider-example-o7k0g). |                                   |
| `enabled`         | `boolean`                           | enables or disables the slider gestures                                                                                                                                                                                                    | `true`                            |
| `vertical`        | `boolean`                           | enables vertical sliding mode                                                                                                                                                                                                              | `false`                           |
| `draggedScale`    | `Number`                            | scale factor of the slides when dragged                                                                                                                                                                                                    | `1.0`                             |
| `draggedSpring`   | `object`                            | spring between the pointer and the dragged slide                                                                                                                                                                                           | `{ tension: 1200, friction: 40 }` |
| `trailingSpring`  | `object`                            | spring of the other slides                                                                                                                                                                                                                 | `{ tension: 120, friction: 30 }`  |
| `releaseSpring`   | `object`                            | spring used when the slides rest (user releases the pointer)                                                                                                                                                                               | `{ tension: 120, friction: 30 }`  |
| `trailingDelay`   | `Number`                            | delay of trailing slides (in ms)                                                                                                                                                                                                           | `50`                              |
| `onDragStart()`   | `(pressedIndex: number) => void`    | function called when the drag starts, passing the index of the slide being dragged as an argument                                                                                                                                          |                                   |
| `onDragEnd()`     | `(pressedIndex: number) => void`    | function called when the drag ends, passing the index of the slide being dragged as an argument                                                                                                                                            |                                   |
| `className`       | `string`                            | CSS class passed to the slider wrapper                                                                                                                                                                                                     |                                   |
| `style`           | `object`                            | style passed to the slider wrapper                                                                                                                                                                                                         |                                   |
| `slideStyle`      | `object` or `(i: number) => object` | style passed to the slides                                                                                                                                                                                                                 |                                   |
| `slideClassName`  | `string`                            | CSS class passed to the slides                                                                                                                                                                                                             |                                   |
| `slideAlign`      | `string (align-items prop)`         | slide alignment (`'center'`, `'flex-start'`, `'flex-end'`)                                                                                                                                                                                 | `'center'`                        |

### Springs configuration

React-soft-slider uses two springs, one for the dragged slide, and one for the other slides, that you can configure to your liking. It accepts any options supported by react-spring, including durations if you're not happy with how springs feel [see here for more info](https://www.react-spring.io/docs/hooks/api).

## Gotchas

**Sizing the slider**

The slider wrapper has a default width set to `100%`, so that it fills its container by default. You can override this behaviour by passing your own `style` or `className` props.

**Sizing your slides relatively to the slider**

If you want to size your slides relatively to the slider width (let's say `width: 80%`), you'll need to rely on `slideStyle` set to `{{ minWidth: '80%' }}` and styling your slide with `width` set to `100%`. The same logic applies for height when in vertical sliding mode.

**Don't use transform styling in slideStyle**

React-soft-slider uses the `transform` attribute to make slides move so transform attributes in `slideStyle` will get overriden.

**React-soft-slider is open to suggestions!**

React-soft-slider will probably never include slider peripheral features, but is open to suggestions to make handling your slides easier!

### Polyfills

You can import the
[IntersectionObserver polyfill](https://www.npmjs.com/package/intersection-observer) and [ResizeObserver polyfill](https://www.npmjs.com/package/resize-observer-polyfill) directly or use
a service like [polyfill.io](https://polyfill.io/v2/docs/) to add it when
needed.

```sh
yarn add intersection-observer resize-observer-polyfill
```

Then import it in your app:

```js
import 'intersection-observer'
import 'resize-observer-polyfill'
```

If you are using Webpack (or similar) you could use
[dynamic imports](https://webpack.js.org/api/module-methods/#import-), to load
the Polyfill only if needed. A basic implementation could look something like
this:

```js
/**
 * Do feature detection, to figure out which polyfills needs to be imported.
 **/
async function loadPolyfills() {
  if (typeof window.IntersectionObserver === 'undefined') {
    await import('intersection-observer')
  }
  if (typeof window.ResizeObserver === 'undefined') {
    await import('resize-observer-polyfill')
  }
}
```


CHANGELOG.md
# Changelog

# 2.2.2 Release

**Summary:** Minor style adjustments.

### Changes

1. Only add `style.width: '100%'` when no `className` prop is provided.

# 2.2.0 Release

**Summary:** Removes outer div, adds className as styling options.

### Changes

1. Removed outer div.
2. Remove `rootStyle` option.

### Added

1. Add `styleClassName` prop to style slides.

### Fixes

1. `slideAlign` behaviour in vertical mode should work properly.

# 2.1.2 Release

**Summary:** Added `indexRange` and `rootStyle` options.

### Added

1. `indexRange: [number, number]` config option to artificially limit the index ranges.
2. `rootStyle` option styles the inner root `<div/>`.

# 2.1.1 Release

**Summary:** This lib is now using `tsdx`.

### Added

1. `releaseSpring` config option.
2. `slideAlign` option to align slides.

### Changes

1. Removed intersection observer polyfill.

# 2.0.0 Release

**Summary:** Move to `react-use-gesture` v7 and `react-spring` v9. Shouldn't break anything but the trailing behavior is slightly different so you might want to keep v1.x.

### Added

1. `trailingDelay` option.

# 1.0.5-7 Releases

**Summary:** Package updates.

# 1.0.4 Release

**Summary:** Removes lodash dependency for `clamp`.

# 1.0.3 Release

**Summary:** Fixed unmount bug affecting IntersectionObserver.

# 1.0.2 Release

**Summary:** Deps update.

## 1.0.1 Release

**Summary:** Added vertical mode.

### Added

1. Added vertical prop to set a vertical sliding mode.


rollup.config.js
import path from 'path'
import pkg from './package.json'
import babel from 'rollup-plugin-babel'
import resolve from 'rollup-plugin-node-resolve'

const extensions = ['.js', '.jsx', '.ts']

const root = process.platform === 'win32' ? path.resolve('/') : '/'
const external = id => !id.startsWith('.') && !id.startsWith(root)

const getBabelOptions = ({ useESModules }, targets) => ({
  babelrc: false,
  comments: false,
  extensions,
  exclude: '**/node_modules/**',
  runtimeHelpers: true,
  presets: [['@babel/preset-env', { loose: true, modules: false, targets }], '@babel/preset-react'],
  plugins: [
    ['@babel/plugin-proposal-object-rest-spread', { loose: true }],
    ['@babel/plugin-transform-runtime', { regenerator: false, useESModules }]
  ]
})

export default [
  {
    input: `./src/index`,
    output: { file: `dist/${pkg.module}`, format: 'esm' },
    external,
    plugins: [babel(getBabelOptions({ useESModules: true }, '>1%, not dead, not ie 11, not op_mini all')), resolve({ extensions })]
  },
  {
    input: `./src/index`,
    output: { file: `dist/${pkg.main}`, format: 'cjs' },
    external,
    plugins: [babel(getBabelOptions({ useESModules: false })), resolve({ extensions })]
  }
]


example/src/dat.js
import React from 'react'
import DatGui, { DatBoolean, DatButton, DatNumber, DatFolder, DatSelect } from '@tim-soft/react-dat-gui'

import { slides, draggedSpringOptions, trailingSpringOptions } from './data'

const Dat = ({ data, onUpdate }) => {
  const addSlide = () => data.nbSlides < slides.length - 1 && onUpdate(({ nbSlides, ...rest }) => ({ ...rest, nbSlides: nbSlides + 1 }))

  return (
    <DatGui data={data} onUpdate={onUpdate}>
      <DatFolder title="Slide appearance">
        <DatBoolean path="variableWidth" label="Variable width" />
        <DatBoolean path="variableHeight" label="Variable height" />
      </DatFolder>
      <DatFolder title="Slides">
        <DatBoolean path="enabled" label="Enabled" />
        <DatBoolean path="autoplay" label="Autoplay" />
        <DatBoolean path="vertical" label="Vertical" />
        <DatNumber path="index" label="Slide index" min={0} max={data.nbSlides - 1} step={1} />
        <DatNumber path="sliderWidth" label="Slider width" min={10} max={100} step={10} />
        <DatButton label={`Add slide (${data.nbSlides})`} onClick={addSlide} />
      </DatFolder>
      <DatFolder title="Configuration">
        <DatBoolean path="trail" label="Trail" />
        <DatNumber path="draggedScale" label="Dragged scale" min={0.1} step={0.1} />
        <DatSelect path="draggedSpring" label="Dragged spring" options={Object.keys(draggedSpringOptions)} />
        <DatSelect path="trailingSpring" label="Trailing spring" options={Object.keys(trailingSpringOptions)} />
      </DatFolder>
    </DatGui>
  )
}

export default Dat


example/src/data.js
import { config } from 'react-spring'

export const slides = [
  'https://images.pexels.com/photos/1055272/pexels-photo-1055272.jpeg?auto=compress&cs=tinysrgb&dpr=2&w=800',
  'https://images.pexels.com/photos/704569/pexels-photo-704569.jpeg?auto=compress&cs=tinysrgb&dpr=2&w=800',
  'https://images.pexels.com/photos/953206/pexels-photo-953206.jpeg?auto=compress&cs=tinysrgb&dpr=2&w=800',
  'https://images.pexels.com/photos/704571/pexels-photo-704571.jpeg?auto=compress&cs=tinysrgb&dpr=2&w=800',
  'https://images.pexels.com/photos/735280/pexels-photo-735280.jpeg?auto=compress&cs=tinysrgb&dpr=2&w=800',
  'https://images.pexels.com/photos/376464/pexels-photo-376464.jpeg?auto=compress&cs=tinysrgb&dpr=2&w=800',
  'https://images.pexels.com/photos/85910/coffee-hot-mug-stripes-85910.jpeg?auto=compress&cs=tinysrgb&dpr=2&w=800',
  'https://images.pexels.com/photos/5510/bread-food-plate-rucola.jpg?auto=compress&cs=tinysrgb&dpr=2&w=800',
  'https://images.pexels.com/photos/543730/pexels-photo-543730.jpeg?auto=compress&cs=tinysrgb&dpr=2&w=800'
]

export const defaultState = {
  enabled: true,
  autoplay: false,
  vertical: false,
  index: 0,
  nbSlides: 4,
  trail: true,
  draggedScale: 0.8,
  sliderWidth: 80,
  variableWidth: false,
  variableHeight: false,
  draggedSpring: 'default',
  trailingSpring: 'default'
}

export const draggedSpringOptions = {
  ...config,
  default: undefined
}

export const trailingSpringOptions = {
  ...config,
  default: undefined
}


example/src/example.js
import React, { useState, useRef, useCallback, useEffect } from 'react'
import { render } from 'react-dom'
import { Slider } from 'react-soft-slider'
import Dat from './dat'
import { defaultState, slides, draggedSpringOptions, trailingSpringOptions } from './data'

import './style.css'

function App() {
  const [state, setState] = useState(defaultState)
  const timeout = useRef()
  const {
    autoplay,
    enabled,
    vertical,
    index,
    nbSlides,
    trail,
    draggedScale,
    sliderWidth,
    variableHeight,
    variableWidth,
    draggedSpring,
    trailingSpring
  } = state

  const setIndex = useCallback(index => setState({ ...state, index }), [state])

  const startAutoplay = useCallback(() => {
    if (autoplay) timeout.current = setInterval(() => setIndex((index + 1) % nbSlides), 5000)
  }, [autoplay, index, nbSlides, setIndex])

  const stopAutoplay = useCallback(() => void clearTimeout(timeout.current), [])

  const handleClick = i => {
    if (i !== state.index) {
      setIndex(i)
    }
  }

  useEffect(() => {
    startAutoplay()
    return stopAutoplay
  }, [startAutoplay, stopAutoplay])

  return (
    <>
      <Dat data={state} onUpdate={setState} />
      <Slider
        enabled={enabled}
        vertical={vertical}
        index={index}
        className="wrapper"
        style={{ width: `${sliderWidth}vw` }}
        slideStyle={vertical ? { minHeight: '100%' } : variableWidth ? undefined : { minWidth: '100%' }}
        onIndexChange={setIndex}
        trail={trail}
        onDragStart={stopAutoplay}
        onDragEnd={startAutoplay}
        draggedScale={draggedScale}
        draggedSpring={draggedSpringOptions[draggedSpring]}
        trailingSpring={trailingSpringOptions[trailingSpring]}
      >
        {slides.slice(0, nbSlides).map((url, i) => (
          <div
            className="slide"
            key={i}
            onClick={() => handleClick(i)}
            style={{
              width: variableWidth ? `${400 + (i % 3) * 50}px` : '100%',
              margin: variableWidth ? `0 ${10 + (i % 5) * 5}px` : '0 10px',
              height: variableHeight ? `${50 + (i % 2) * 10}%` : '80%',
              backgroundImage: `url(${url})`
            }}
          />
        ))}
      </Slider>
    </>
  )
}

render(<App />, document.getElementById('root'))


src/index.js
import React, { useRef, useEffect } from 'react'
import PropTypes from 'prop-types'
import { useSprings, animated } from 'react-spring'
import { useDrag } from 'react-use-gesture'
import useResizeObserver from 'use-resize-observer'
import 'intersection-observer'

// style for the slides wrapper
const slidesWrapperStyle = vertical => ({
  display: 'flex',
  height: '100%',
  width: '100%',
  flexWrap: 'nowrap',
  alignItems: 'stretch',
  position: 'relative',
  WebkitUserSelect: 'none',
  userSelect: 'none',
  WebkitTouchCallout: 'none',
  flexDirection: vertical ? 'column' : 'row',
  touchAction: vertical ? 'pan-x' : 'pan-y'
})

const clamp = (num, clamp, higher) => Math.min(Math.max(num, clamp), higher)

export function Slider({
  children,
  index,
  onIndexChange,
  className,
  style,
  slideStyle,
  enabled,
  trail,
  vertical,
  draggedScale,
  draggedSpring,
  trailingSpring,
  trailingDelay,
  onDragStart,
  onDragEnd
}) {
  const slideStyleFunc = typeof slideStyle === 'function' ? slideStyle : () => slideStyle
  // root holds are slides wrapper node and we use a ResizeObserver
  // to observe its size in order to recompute the slides position
  // when it changes
  const root = useRef()
  const { width, height } = useResizeObserver({ ref: root })

  const axis = vertical ? 'y' : 'x'
  const size = vertical ? height : width

  // indexRef is an internal reference to the current slide index
  const indexRef = useRef(index)

  // restPos holds a reference to the adjusted position of the slider
  // when rested
  const restPos = useRef(0)
  const velocity = useRef(0)

  // visibleIndexes is a Set holding the index of slides that are
  // currently partially or fully visible (intersecting) in the
  // viewport
  const visibleIndexes = useRef(new Set())
  const firstVisibleIndex = useRef(0)
  const lastVisibleIndex = useRef(0)

  // instances holds a ref to an array of controllers
  // to simulate a spring trail. Mechanics is directly
  // copied from here https://github.com/react-spring/react-spring/blob/31200a79843ce85200b2a7692e8f14788e60f9e9/src/useTrail.js#L14
  // const instances = useRef()

  // callback called by the intersection observer updating
  // visibleIndexes
  const cb = slides => {
    slides.forEach(({ isIntersecting, target }) =>
      visibleIndexes.current[isIntersecting ? 'add' : 'delete'](~~target.attributes['data-index'].value)
    )
    const visibles = Array.from(visibleIndexes.current).sort()
    firstVisibleIndex.current = visibles[0]
    lastVisibleIndex.current = visibles[visibles.length - 1]
  }

  const observer = useRef(null)

  // we add the slides to the IntersectionObserver:
  // this is recomputed everytime the user adds or removes a slide
  useEffect(() => {
    if (!observer.current) observer.current = new IntersectionObserver(cb)
    Array.from(root.current.children).forEach(t => observer.current.observe(t))
  }, [children.length, root])

  // removing the observer on unmount
  useEffect(() => () => observer.current.disconnect(), [])

  // setting the springs with initial position set to restPos:
  // this is important when adding slides since changing children
  // length recomputes useSprings
  const [springs, set] = useSprings(
    children.length,
    i => {
      // zIndex will make sure the dragged slide stays on top of the others
      return {
        x: vertical ? 0 : restPos.current,
        y: vertical ? restPos.current : 0,
        s: 1,
        zIndex: 0,
        immediate: key => key === 'zIndex'
      }
    },
    [children.length]
  )

  // everytime the index changes, we should calculate the right position
  // of the slide so that its centered: this is recomputed everytime
  // the index changes
  useEffect(() => {
    // if width and height haven't been set don't do anything
    // (this happens on first render before useResizeObserver had the time to kick in)
    if (!width || !height) return
    // here we take the selected slide
    // and calculate its position so its centered in the slides wrapper
    if (vertical) {
      const { offsetTop, offsetHeight } = root.current.children[index]
      restPos.current = Math.round(-offsetTop + (height - offsetHeight) / 2)
    } else {
      const { offsetLeft, offsetWidth } = root.current.children[index]
      restPos.current = Math.round(-offsetLeft + (width - offsetWidth) / 2)
    }
    // two options then:
    // 1. the index was changed through gestures: in that case indexRef
    // is equal to index, we just want to set the position where it should

    if (indexRef.current === index) {
      set(() => ({ [axis]: restPos.current, s: 1, config: key => key === axis && { ...trailingSpring, velocity: velocity.current } }))
    } else {
      // 2. the user has changed the index props: in that case indexRef
      // is outdated and different from index. We want to animate depending
      // on the direction of the slide, with the furthest slide moving first
      // trailing the others

      const dir = index < indexRef.current ? -1 : 1
      // if direction is 1 then the first slide to animate should be the lowest
      // indexed visible slide, if -1 the highest
      const firstToMove = dir > 0 ? firstVisibleIndex.current : lastVisibleIndex.current
      set(i => {
        return {
          [axis]: restPos.current,
          s: 1,
          config: key => key === axis && trailingSpring,
          delay: i * dir < firstToMove * dir ? 0 : Math.abs(firstToMove - i) * trailingDelay
        }
      })
    }
    // finally we update indexRef to match index
    indexRef.current = index
  }, [index, set, root, vertical, axis, height, width, trail, trailingSpring, draggedSpring, trailingDelay])

  // adding the bind listener
  const bind = useDrag(
    ({
      first,
      last,
      vxvy: [vx, vy],
      delta: [dx, dy],
      movement: [movX, movY],
      args: [pressedIndex],
      memo = springs[pressedIndex][axis].getValue()
    }) => {
      const v = vertical ? vy : vx
      const dir = -Math.sign(vertical ? dy : dx)
      const mov = vertical ? movY : movX
      if (first) {
        // if this is the first drag event, we're trailing the controllers
        // to the index being dragged and setting zIndex, scale and config
        set(i => {
          return {
            [axis]: memo + mov,
            s: draggedScale,
            config: key => (key === axis && i === pressedIndex ? draggedSpring : trailingSpring),
            zIndex: i === pressedIndex ? 10 : 0
          }
        })

        // triggering onDragStart prop if it exists
        onDragStart && onDragStart(pressedIndex)
      } else if (last) {
        // when the user releases the drag and the distance or speed are superior to a threshold
        // we update the indexRef
        if (Math.abs(mov) > size / 2 || Math.abs(v) > 0.3) {
          indexRef.current = clamp(indexRef.current + (mov > 0 ? -1 : 1), 0, children.length - 1)
        }
        // if the index is not equal to indexRef we know we've moved a slide
        // so we tell the user to update its index in the next tick and our useEffect
        // will do the rest. RAF is used to make sure we're not updating the index
        // too fast: that might happen if the user wants to update a slide onClick
        // TODO - need an example
        if (index !== indexRef.current) {
          velocity.current = v
          requestAnimationFrame(() => onIndexChange(indexRef.current))
        }
        // if the index hasn't changed then we set the position back to where it should be
        else set(() => ({ [axis]: restPos.current, s: 1, config: key => key === axis && trailingSpring }))

        // triggering onDragEnd prop if it exists
        onDragEnd && onDragEnd(pressedIndex)
      }

      // if not we're just dragging and we're just updating the position
      else {
        const firstToMove = dir > 0 ? firstVisibleIndex.current : lastVisibleIndex.current
        set(i => {
          return {
            [axis]: mov + memo,
            delay: i * dir < firstToMove * dir || i === pressedIndex ? 0 : Math.abs(firstToMove - i) * trailingDelay,
            config: key => (key === axis && i === pressedIndex ? draggedSpring : trailingSpring)
          }
        })
      }

      // and returning memo to keep the initial position in cache along drag
      return memo
    },
    { enabled, axis }
  )

  return (
    <div className={className} style={style}>
      <div ref={root} style={slidesWrapperStyle(vertical)}>
        {springs.map(({ [axis]: pos, s, zIndex }, i) => (
          <animated.div
            // passing the index as an argument will let our handler know
            // which slide is being dragged
            {...bind(i)}
            key={i}
            data-index={i}
            style={{
              ...slideStyleFunc(i),
              zIndex,
              [axis]: pos,
              scale: s,
              display: 'flex',
              alignItems: 'center',
              willChange: 'transform'
            }}
          >
            {children[i]}
          </animated.div>
        ))}
      </div>
    </div>
  )
}

Slider.defaultProps = {
  className: undefined,
  style: undefined,
  enabled: true,
  trail: true,
  vertical: false,
  draggedScale: 1,
  draggedSpring: { tension: 1200, friction: 40 },
  trailingSpring: { mass: 1, tension: 280, friction: 60 },
  trailingDelay: 50,
  onDragStart: undefined,
  onDragEnd: undefined
}

Slider.propTypes = {
  children: PropTypes.node.isRequired,
  index: PropTypes.number.isRequired,
  onIndexChange: PropTypes.func.isRequired,
  className: PropTypes.string,
  style: PropTypes.object,
  slideStyle: PropTypes.object,
  enabled: PropTypes.bool,
  vertical: PropTypes.bool,
  trail: PropTypes.bool,
  draggedScale: PropTypes.number,
  draggedSpring: PropTypes.object,
  trailingSpring: PropTypes.object,
  onDragStart: PropTypes.func,
  onDragEnd: PropTypes.func
}
